Adding pure advection solver

ssafaei opened the pull request:
Advection solver is the fourth solver in 1D transient example
Also the code has been modified to work in parallel mode
Unused variables have been removed

chrispbradley commented on src/advection_equation_routines.f90:
```diff
@@ -0,0 +1,1156 @@
+                CALL FIELD_SCALING_TYPE_SET(EQUATIONS_INDEPENDENT%INDEPENDENT_FIELD,GEOMETRIC_SCALING_TYPE,ERR,ERROR,*999)
+              ELSE
+                INDEPENDENT_FIELD_NUMBER_OF_VARIABLES=3
+                NUMBER_OF_INDEPENDENT_COMPONENTS=2
```
Why do we have 3 variables with 2 components for the advection field? Shouldn't there be 1 variable with number of dimensions components?

ssafaei replied:
It's a common field between navier-stokes and advection solver, I was using the 1st component for the advection velocity and 2nd one for the diffusivity.

chrispbradley replied:
Diffusivity? Shouldn't that be a material field component? Also, by 1st component above do you mean first variable???

ssafaei replied:
Yes, I have to change it to material field. These are the 3 variables and 2 components for each.
IndependentFieldNavierStokes.VariableLabelSet(CMISS.FieldVariableTypes.U,'Normal Wave Direction')
IndependentFieldNavierStokes.VariableLabelSet(CMISS.FieldVariableTypes.V,'Input Boundary Condition')
IndependentFieldNavierStokes.VariableLabelSet(CMISS.FieldVariableTypes.U1,'Advection Velocity')
IndependentFieldNavierStokes.ComponentMeshComponentSet(CMISS.FieldVariableTypes.U,1,MeshComponentNumberSpace)
IndependentFieldNavierStokes.ComponentMeshComponentSet(CMISS.FieldVariableTypes.U,2,MeshComponentNumberSpace)
IndependentFieldNavierStokes.ComponentMeshComponentSet(CMISS.FieldVariableTypes.V,1,MeshComponentNumberSpace)
IndependentFieldNavierStokes.ComponentMeshComponentSet(CMISS.FieldVariableTypes.U1,1,MeshComponentNumberSpace)
IndependentFieldNavierStokes.ComponentMeshComponentSet(CMISS.FieldVariableTypes.U1,2,MeshComponentNumberSpace)

chrispbradley replied:
I think you are confusing the independent field for the Navier Stokes equations set with the independent field for the advection equations set, the fields do not depend on each other (even when coupled). The independent field for the advection problem specifies the advection field. For you coupled problem this will be your Navier Stokes dependent field. For the auto-created field it should be one variable with number of components given by the number of dimensions for the problem.

ssafaei replied:
it's done.

chrispbradley commented on src/advection_equation_routines.f90:
```diff
@@ -0,0 +1,1156 @@
+                !Finish creating the independent field
+                CALL FIELD_CREATE_FINISH(EQUATIONS_INDEPENDENT%INDEPENDENT_FIELD,ERR,ERROR,*999)
+                !Set the default values for the independent field
+                CALL FIELD_PARAMETER_SET_CREATE(EQUATIONS_SET%INDEPENDENT%INDEPENDENT_FIELD,FIELD_U_VARIABLE_TYPE, &
```
Why do we have a field_boundary_set_type parameter set for the independent field?

dladd commented on src/advection_equation_routines.f90:
```diff
@@ -0,0 +1,1269 @@
+ 
+    IF(ASSOCIATED(SOLVER)) THEN
+      SOLVERS=>SOLVER%SOLVERS
+      IF(ASSOCIATED(SOLVERS)) THEN
```
Setting the dependent field conc values to a fixed value here doesn't seem right to me. But the only call for this subroutine appears to be commented out so I guess it's not an issue. Could maybe either remove this subroutine or put a '! do nothing' in and keep the routine as a place holder for the future.

ssafaei replied:
This subroutine is for applying time varying BC on advection-diffusion equation, similar to the one we have in navier-stokes, but because at the moment I was using initial BC, I commented out it.

chrispbradley replied:
As a point of comment, I'm not so sure time varying BCs are a solver thing and thus they shouldn't be changed in a pre-solve. I would have thought at the beginning of the (time) control loop?

dladd commented on src/control_loop_routines.f90:
```diff
@@ -224,29 +224,41 @@ END SUBROUTINE CONTROL_LOOP_CREATE_START
-            TIME_INCREMENT=TIME_LOOP%TIME_INCREMENT
+        CONTROL_LOOP_LEVEL=>CONTROL_LOOP%CONTROL_LOOP_LEVEL
+        PARENT_LOOP=>CONTROL_LOOP
+        DO I=CONTROL_LOOP_LEVEL,1,-1
```
Might just want to check with @chrispbradley whether 0-level control loops can ever be time-loops (this currently assumes they can not)

chrispbradley replied:
I don't see why they can't be time loops?

dladd commented on src/control_loop_routines.f90:
```diff
@@ -1334,38 +1346,55 @@ END SUBROUTINE CONTROL_LOOP_TIME_INITIALISE
-            CURRENT_TIME=TIME_LOOP%CURRENT_TIME
+        CONTROL_LOOP_LEVEL=>CONTROL_LOOP%CONTROL_LOOP_LEVEL
+        PARENT_LOOP=>CONTROL_LOOP
+        DO I=CONTROL_LOOP_LEVEL,1,-1
```
as above

dladd commented on src/opencmiss.f90:
```diff
@@ -16946,6 +16946,8 @@ SUBROUTINE CMISSControlLoop_TimesGetNumber0(problemUserNumber,controlLoopIdentif
     REAL(DP), INTENT(OUT) :: stopTime !<On return, the stop time for the time control loop.
     REAL(DP), INTENT(OUT) :: timeIncrement !<On return, the time increment for the time control loop.
     REAL(DP), INTENT(OUT) :: currentTime !<On return, the current time for the time control loop.
+    INTEGER(INTG), INTENT(OUT) :: currentLoopIteration !<On return, the time increment for the time control loop.
```
The comments for these iteration variables should be updated
e.g.,
!<On return, the current iteration number for the time control loop.

ssafaei committed:
adding veins and heart

ssafaei committed:
Change RCR model

ssafaei committed:
Merge branch 'master' of https://github.com/OpenCMISS/cm

Conflicts:
	src/Navier_Stokes_equations_routines.f90

ssafaei committed:
Update Navier_Stokes_equations_routines.f90

Fix the input flow for Olufsen model

ssafaei committed:
Update Navier_Stokes_equations_routines.f90

Fix the input flow for Olufsen model

ssafaei committed:
Fix the input flow for Olufsen model

ssafaei committed:
Fix a bug

ssafaei committed:
Final commit

ssafaei committed:
Fix conflict
Merge branch 'master' of https://github.com/OpenCMISS/cm

Conflicts:
	src/field_routines.f90
	src/opencmiss.f90

ssafaei committed:
Merge branch 'master' of https://github.com/Soroosh/cm

Conflicts:
	src/Navier_Stokes_equations_routines.f90

ssafaei committed:
Final modification for 1D/0D coupling

ssafaei committed:
Adding structured tree code

Commit:
Commit structured tree code - final version

Commit:
Merge branch 'master' of https://github.com/OpenCMISS/cm

Commit:
Adding heart and G to Navier-Stokes
A minor bug in python bindings
Few changes in FSI problem

chrispbradley commented:
@dladd @ssafaei Is this pull request still valid or has it been dealt with in the big fluids pull request from David?

dladd commented:
@ssafaei @chrispbradley  It looks to me like everything in here was folded into the other request- for example the structured tree and heart and gravity commits all seem to be there. @ssafaei do you see anything else or have a test case for the functionality from this pull request? Cheers

ssafaei commented:
I think it's all good, it's folded into the other request.

ssafaei committed:
Adding the venous system

ssafaei committed:
Merge branch 'master' of https://github.com/OpenCMISS/cm

Conflicts:
	src/Navier_Stokes_equations_routines.f90
	src/advection_equation_routines.f90
	src/boundary_condition_routines.f90
	src/characteristic_equation_routines.f90
	src/cmiss_mpi.f90
	src/cmiss_parmetis.f90
	src/cmiss_petsc.f90
	src/control_loop_routines.f90
	src/distributed_matrix_vector.f90
	src/domain_mappings.f90
	src/elasticity_routines.f90
	src/equations_routines.f90
	src/equations_set_constants.f90
	src/equations_set_routines.f90
	src/field_IO_routines.f90
	src/field_routines.f90
	src/finite_elasticity_routines.f90
	src/fluid_mechanics_routines.f90
	src/fsi_routines.f90
	src/linkedlist_routines.f90
	src/lists.f90
	src/maths.f90
	src/matrix_vector.f90
	src/mesh_routines.f90
	src/opencmiss.f90
	src/problem_constants.f90
	src/solver_routines.f90
	src/stree_equation_routines.f90

ssafaei committed:
Add linearised terms in P-A

ssafaei committed:
latest commit

ssafaei committed:
Merge branch 'master' of https://github.com/OpenCMISS/cm

ssafaei committed:
Merge branch 'master' of https://github.com/ssafaei/cm
